### ISSUE TITLE
Add gifs of usage to widget examples

### ISSUE THREAD

**Issue by ianhi** [matplotlib/matplotlib#19222]
### Problem

It can be hard to discern what a widget example does as for many of them there are no images that show up in the docs. For example the [Lasso Selector Demo](https://matplotlib.org/devdocs/gallery/widgets/lasso_selector_demo_sgskip.html#sphx-glr-gallery-widgets-lasso-selector-demo-sgskip-py) has no image and the [Lasso Demo](https://matplotlib.org/devdocs/gallery/event_handling/lasso_demo.html#sphx-glr-gallery-event-handling-lasso-demo-py) only has a static image. 

### Suggested Improvement
Manually generate GIFS of the usage and embed them onto the example pages. For example here is a GIF of me using the LassoSelector demo:
![lasso-selector](https://user-images.githubusercontent.com/10111092/103472284-c2c43a00-4d59-11eb-8c03-59ca6e2b9489.gif)

I'm not sure if sphinx-gallery supports this though? maybe embeding and html `<img>` tag into the comment at the top would work?
Also instead of `.gifs` I it would be better to keep files small by using `.apng` and running them https://tinypng.com/

**Comment by story645** [comment]
Agree, and also #17479 has a discussion on the viability of doing something automated.

**Comment by timhoffm** [comment]
The animation is a useful addition to make the widget examples more easy to understand.

This is not particular difficult but somewhat tedious because AFAICS it will need some manual steps and explicit configuration.

### Automated creation:
While it may be possible to automatically generate a sceenshot from static widgets. This animation here requires user interaction, and I assume it's not worth the effort to script user interaction for the examples.

### Manually generate GIFS
I'm happy to see it if somebody takes the responsibility.

What would have to be done:
1. Generate a useful animated image (gif/apng).
2. Embed that explicitly into the example using sphinx `.. image::` directive.
3. Create a downsampled version of that image for the tumbnail and configure sphinx-gallery to use this explicitly](https://sphinx-gallery.github.io/stable/configuration.html#providing-an-image-for-the-thumbnail-image)

The whole process (in particular also the tooling used in 1.) should be described in the [documentation guide](https://matplotlib.org/devdocs/devel/documenting_mpl.html) so that we and others are able to reproduce it if necessary.

**Comment by QuLogic** [comment]
We _do_ have a way of [mocking events to widgets](https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/testing/widgets.py), so it might be possible to write some automated way to generate these. It would still require some work somewhere though to implement it all.

**Comment by timhoffm** [comment]
> We _do_ have a way of mocking events to widgets

The downside is that you don't get the cursor. IMHO seeing the cursor movement significantly helps in understanding what's going on.

**Comment by ianhi** [comment]
I agree that seeing the cursor is crucial for understanding, however manually recording the GIFs is also sad times because it would need to be redone anytime the styling changes. I think we can get the best of both worlds by doing:

1. Mock the events and save these in a python file somewhere
   - I think this can be easily done by interacting by hand and just printing out the events and then throwing them into a python file.
2. At docs build time run through the mock events + manually draw in a shape that looks like a mouse cursor
3. Save that as an animation somewhere - can just be a FuncAnimation or something.
4. Use a sphinx-gallery comment to embed the generated gif

This also carries the extra advantage of not increasing the repo size as the only thing that is persistent is a text file of the mouseevents.

**Comment by timhoffm** [comment]
Um, this is doable but sounds like a medium size effort. I'm somewhat reluctant taking this on in core Matplotlib. 1-3 could be developed as a 3rd party package.

**Comment by jklymak** [comment]
Agreed - this is a lot of machinery for something the user could just try themselves from the example.  

The low-tech way of doing this is to do them by hand, and then putting a comment in the widget code noting that they should be modified if there are major changes.  That is pretty low maintenance and it is not like widgets are changing very often.

**Comment by ianhi** [comment]
> Um, this is doable but sounds like a medium size effort. I'm somewhat reluctant taking this on in core Matplotlib. 1-3 could be developed as a 3rd party package.

I think it makes sense as a third party package. I would use this, and I think animatplotlib would also find it useful. To that end I couldn't get it out of my head so I did some maybe terrible things using `exec` and made this https://github.com/ianhi/mpl-playback

Which I used to generate this gif:
![](https://raw.githubusercontent.com/ianhi/mpl-playback/main/played-back.gif)

**Comment by ianhi** [comment]
But I hear the arguments about not adding extra machinery. If matplotlib ends up going the route of manually recording gifs I won't be made sad :)

**Comment by story645** [comment]
Is there also the option of making it a 3rd party package that is then a dependency in the doc builds?

**Comment by QuLogic** [comment]
> * Mock the events and save these in a python file somewhere
>   
>   * I think this can be easily done by interacting by hand and just printing out the events and then throwing them into a python file.
> 
> * At docs build time run through the mock events + manually draw in a shape that looks like a mouse cursor
> 
> * Save that as an animation somewhere - can just be a FuncAnimation or something.
> 
> * Use a sphinx-gallery comment to embed the generated gif

This is essentially what I would do, with some small tweaks. Sphinx-gallery has a [custom scraper API](https://sphinx-gallery.github.io/stable/advanced.html#write-a-custom-image-scraper), so I would use that instead to create and inject the resulting GIF.

**Comment by ianhi** [comment]
Thanks for the pointer @QuLogic, that does seem to be the correct approach. I don't think I handled the thumbnail stuff correctly yet, but here is a working gallery with GIFs:
https://mpl-playback.readthedocs.io/en/latest/auto_examples/index.html
